### PR TITLE
GPX - BLE Graphs

### PR DESCRIPTION
**Pull request by Skalii**

### CODE REVIEW

File: Sources/GPX/OAGPXTrackAnalysis.mm [hunk 1]
```diff
@@ -259,6 +260,19 @@ -(BOOL) isSpeedSpecified
     return _avgSpeed > 0.0;
 }
 
+- (BOOL)hasData:(NSString *)tag
+{
+    return [_availableAttributes containsObject:tag];
+}
+
+- (void)setHasData:(NSString *)tag hasData:(BOOL)hasData
```
Comment (tigrim) [inline Sources/GPX/OAGPXTrackAnalysis.mm:268]: - (void)updateAvailabilityForAttributeTag:(NSString *)tag hasData:(BOOL)hasData

File: Sources/Helpers/Extensions/ExternalSensorsPlugin+Ext.swift [hunk 2]
```diff
@@ -0,0 +1,22 @@
+//
+//  ExternalSensorsPlugin+Ext.swift
```
Comment (tigrim) [inline Sources/Helpers/Extensions/ExternalSensorsPlugin+Ext.swift:2]: ExternalSensorsPlugin+Extension

File: Sources/Helpers/OAAppSettings.h [hunk 3]
```diff
@@ -13,6 +13,7 @@
 #import "OAColoringType.h"
 #import "OADownloadMode.h"
 #import "OAMap3DModeVisibilityType.h"
+//#import "OsmAnd_Maps-Swift.h"
```
Comment (tigrim) [inline Sources/Helpers/OAAppSettings.h:16]: remove

File: Sources/Plugins/ExternalSensorsPlugin/SensorAttributesUtils.swift [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    @objc static func getOrderedLineDataSet(chart: LineChartView,
+                                                   analysis: OAGPXTrackAnalysis,
+                                                   graphType: GPXDataSetType,
+                                                   axisType: GPXDataSetAxisType,
+                                                   calcWithoutGaps: Bool,
+                                                   useRightAxis: Bool) -> GpxUIHelper.OrderedLineDataSet? {
+        switch graphType {
+        case .SENSOR_SPEED:
```
Comment (tigrim) [inline Sources/Plugins/ExternalSensorsPlugin/SensorAttributesUtils.swift:65]: case .SENSOR_SPEED where hasSensorSpeedData(analysis):

File: Sources/Plugins/OAPlugin.mm [hunk 5]
```diff
@@ -957,4 +957,30 @@ + (void)enablePluginsByMapWidgets:(NSSet<NSString *> *)widgetIds
...
+{
+    for (OAPlugin *plugin in [self getEnabledPlugins])
+    {
+        [plugin onAnalysePoint:analysis point:point attribute:attribute];
+    }
+}
+
+- (void)onAnalysePoint:(OAGPXTrackAnalysis *)analysis point:(NSObject *)point attribute:(OAPointAttributes *)attribute
+{
+}
```
Comment (tigrim) [inline Sources/Plugins/OAPlugin.mm:970]: empty?

File: Sources/Controllers/BottomSheet/RouteInfoBottomSheet/OAStatisticsSelectionBottomSheetViewController.mm [hunk 6]
```diff
@@ -77,50 +76,47 @@ - (void) setupView
...
-    [arr addObject:@{
-        @"type" : [OATitleIconRoundCell getCellIdentifier],
-        @"title" : OALocalizedString(@"shared_string_slope"),
-        @"img" : @"ic_custom_ascent",
-        @"mode" : @(EOARouteStatisticsModeSlope),
-        @"round_bottom" : @(NO),
-        @"round_top" : @(NO)
-    }];
+    NSMutableArray<NSArray<NSNumber *> *> *allTypes = [NSMutableArray array];
+    [allTypes addObject:@[@(GPXDataSetTypeALTITUDE)]];
```
Comment (tigrim) [inline Sources/Controllers/BottomSheet/RouteInfoBottomSheet/OAStatisticsSelectionBottomSheetViewController.mm:80]: the best for enum use camelcase style: case GPXDataSetTypeAltitude,

File: Sources/Controllers/RouteInfoMenu/GpxUIHelper.swift [hunk 7]
```diff
@@ -10,64 +10,64 @@ import UIKit
...
     public func getName() -> String {
         switch self {
-        case .DISTANCE:
+        case .distance:
             return OAUtilities.getLocalizedString("shared_string_distance");
-        case .TIME:
+        case .time:
             return OAUtilities.getLocalizedString("shared_string_time");
-        case .TIMEOFDAY:
+        case .timeofday:
```
Comment (tigrim) [inline Sources/Controllers/RouteInfoMenu/GpxUIHelper.swift:59]: timeOfDay - camelCase style

**Review by tigrim — APPROVED** [review]

File: Sources/Controllers/RouteInfoMenu/GpxUIHelper.swift [hunk 8]
```diff
@@ -146,8 +154,8 @@ import Charts
             return dateFormatter.string(from: date)
         }
     }
-    
-    private class OrderedLineDataSet: LineChartDataSet {
+
+    public class OrderedLineDataSet: LineChartDataSet {
```
Comment (tigrim) [inline Sources/Controllers/RouteInfoMenu/GpxUIHelper.swift:158]: final class (remove public)

File: Sources/Controllers/RouteInfoMenu/GpxUIHelper.swift [hunk 9]
```diff
@@ -75,7 +83,7 @@ import Charts
     
     private static let MAX_CHART_DATA_ITEMS: Double = 10000
     
-    private class ValueFormatter: IAxisValueFormatter
+    public class ValueFormatter: IAxisValueFormatter
```
Comment (tigrim) [inline Sources/Controllers/RouteInfoMenu/GpxUIHelper.swift:86]: remove  public

File: Sources/GPX/OAGPXTrackAnalysis.mm (hunk 1 above)
Comment (tigrim) [reply Sources/GPX/OAGPXTrackAnalysis.mm:268]: fix

File: Sources/Plugins/OAPlugin.mm (hunk 5 above)
Comment (tigrim) [reply Sources/Plugins/OAPlugin.mm:970]: ?

File: Sources/Controllers/RouteInfoMenu/GpxUIHelper.swift [hunk 10]
```diff
@@ -10,56 +10,64 @@ import UIKit
 import Charts
 
 @objc public enum GPXDataSetType: Int {
-    case ALTITUDE = 0
-    case SPEED = 1
-    case SLOPE = 2
+    case altitude = 0
```
Comment (tigrim) [inline Sources/Controllers/RouteInfoMenu/GpxUIHelper.swift:13]: we can use it without default values, just` case altitude, speed`

File: Sources/Controllers/RouteInfoMenu/GpxUIHelper.swift [hunk 11]
```diff
@@ -10,56 +10,64 @@ import UIKit
...
+        OAGPXDataSetType.getFillColor(self.rawValue)
+    }
+
+    public func getMainUnitY() -> String {
+        OAGPXDataSetType.getMainUnitY(self.rawValue)
     }
 }
 
 @objc public enum GPXDataSetAxisType: Int {
-    case DISTANCE = 0
```
Comment (tigrim) [inline Sources/Controllers/RouteInfoMenu/GpxUIHelper.swift:41]: we can use it without default values

File: Sources/Plugins/ExternalSensorsPlugin/SensorAttributesUtils.swift [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    @objc static func getOrderedLineDataSet(chart: LineChartView,
+                                                   analysis: OAGPXTrackAnalysis,
+                                                   graphType: GPXDataSetType,
+                                                   axisType: GPXDataSetAxisType,
+                                                   calcWithoutGaps: Bool,
+                                                   useRightAxis: Bool) -> GpxUIHelper.OrderedLineDataSet? {
+        switch graphType {
+        case .sensorSpeed:
```
Comment (tigrim) [inline Sources/Plugins/ExternalSensorsPlugin/SensorAttributesUtils.swift:65]: case .sensorSpeed where hasSensorSpeedData(analysis) :

**Review by tigrim — APPROVED** [review]

File: Sources/Helpers/OAAppSettings.h [hunk 13]
```diff
@@ -61,6 +61,17 @@ typedef NS_ENUM(NSInteger, EOACompassMode)
 
 @end
 
+@interface OAGPXDataSetType : NSObject
```
Comment (alex-osm) [inline Sources/Helpers/OAAppSettings.h:64]: Why this class in settings?

**Pull request merged**